python/perf-kernels/streamk/streamk_kernel.py

```diff
@@ -0,0 +1,221 @@
+        total_streamk_tiles = total_tiles % num_sms
+        total_full_tiles = total_tiles - total_streamk_tiles
+        total_streamk_iters = total_streamk_tiles * iters_per_tile
+        # iterations related to full waves
+        streamk_iters_pcu = total_streamk_iters // num_sms
+        # iterations related to last (partial) wave
+        streamk_remainder_iters = total_streamk_iters % num_sms
+
+    else:  # all tiles are computed using classical blocking
+        total_full_tiles = total_tiles
+        total_streamk_tiles = 0
+        streamk_iters_pcu = 0
+        streamk_remainder_iters = 0
+        total_streamk_iters = 0
+
+    return iters_per_tile, total_full_tiles, total_streamk_tiles, streamk_iters_pcu, streamk_remainder_iters
+
+
+@triton.jit()
+def streamk_gemm(
+    A,
+    B,
+    C,
+    bias_ptr,
+    P,
+    locks,
+    M,
+    N,
+    K,
+    num_sms,
+    stride_am,
+    stride_ak,
+    stride_bk,
+    stride_bn,
+    stride_cm,
+    stride_cn,
+    stride_bias,
+    BLOCK_SIZE_M: tl.constexpr,
+    BLOCK_SIZE_N: tl.constexpr,
+    BLOCK_SIZE_K: tl.constexpr,
+    GROUP_SIZE_M: tl.constexpr,
+    BIAS: tl.constexpr,
+    EVEN_K: tl.constexpr,
+):
+    pid = tl.program_id(0)
+    pid = get_new_pid(pid, num_sms)
+    num_pid_m = tl.cdiv(M, BLOCK_SIZE_M)
+    num_pid_n = tl.cdiv(N, BLOCK_SIZE_N)
+
+    iters_per_tile, total_full_tiles, total_streamk_tiles, streamk_iters_pcu, streamk_remainder_iters = get_tiles_config(
+        M, N, K, num_sms, BLOCK_SIZE_M, BLOCK_SIZE_N, BLOCK_SIZE_K)
+
+    acc_dtype = tl.float32 if C.type.element_ty != tl.int8 else tl.int32
+    rk = tl.arange(0, BLOCK_SIZE_K)
+
+    for tile_id in range(pid, total_full_tiles, num_sms):
+        if GROUP_SIZE_M == 1:
+            pid_m = tile_id // num_pid_n
+            pid_n = tile_id % num_pid_n
+        else:
+            num_pid_in_group = GROUP_SIZE_M * num_pid_n
+            group_id = tile_id // num_pid_in_group
+            first_pid_m = group_id * GROUP_SIZE_M
+            group_size_m = min(num_pid_m - first_pid_m, GROUP_SIZE_M)
+            pid_m = first_pid_m + (tile_id % group_size_m)
+            pid_n = (tile_id % num_pid_in_group) // group_size_m
+
+        rm = (pid_m * BLOCK_SIZE_M + tl.arange(0, BLOCK_SIZE_M)) % M
+        rn = (pid_n * BLOCK_SIZE_N + tl.arange(0, BLOCK_SIZE_N)) % N
+        rm = tl.max_contiguous(tl.multiple_of(rm, BLOCK_SIZE_M), BLOCK_SIZE_M)
+        rn = tl.max_contiguous(tl.multiple_of(rn, BLOCK_SIZE_N), BLOCK_SIZE_N)
+        A_BASE = A + rm[:, None] * stride_am + rk[None, :] * stride_ak
+        B_BASE = B + rk[:, None] * stride_bk + rn[None, :] * stride_bn
+
+        if BIAS:
+            bias_ = bias_ptr + rm * stride_bias
+            bias = tl.load(bias_, mask=rm < M, other=0.0)
+
+        acc = tl.zeros((BLOCK_SIZE_M, BLOCK_SIZE_N), dtype=acc_dtype)
+        for k in range(0, tl.cdiv(K, BLOCK_SIZE_K)):
+            if EVEN_K:
+                a = tl.load(A_BASE)
```

Comment:
Can we peel the masking for the last iteration when `EVEN_K is False` so that only the last loop pays the price of the mask?

Reply:
as discussed, this will be in next PR. Thanks !